openassessment/fileupload/views.py

```diff
@@ -0,0 +1,151 @@
+import hashlib
+import json
+import os
+
+from django.conf import settings
+from django.shortcuts import HttpResponse, Http404
+from django.utils import timezone
+from django.views.decorators.http import require_http_methods
+
+from . import exceptions
+from .backends.filesystem import is_upload_url_available, is_download_url_available
+from .backends.base import Settings
+
+
+@require_http_methods(["PUT", "GET"])
+def filesystem_storage(request, key):
+    """
+    Uploading and download files to the local filesystem backend.
+    """
+    if isinstance(key, unicode):
+        key = key.encode("utf-8")
+    if request.method == "PUT":
+        if not is_upload_url_available(key):
+            raise Http404()
+        content, metadata = get_content_metadata(request)
+        save_to_file(key, content, metadata)
+        return HttpResponse()
+    elif request.method == "GET":
+        if not is_download_url_available(key):
+            raise Http404()
+        return download_file(key)
+
+
+def download_file(key):
+    """Returns an HttpResponse to download the corresponding file"""
+
+    file_path = get_file_path(key)
+    metadata_path = get_metadata_path(key)
+    if not os.path.exists(file_path):
+        raise Http404()
+    with open(metadata_path) as f:
+        metadata = json.load(f)
+        content_type = metadata.get("Content-Type", 'application/octet-stream')
+    with open(file_path, 'r') as f:
+        response = HttpResponse(f.read(), content_type=content_type)
+        file_name = os.path.basename(os.path.dirname(file_path))
+        response['Content-Disposition'] = 'attachment; filename=' + file_name
+    return response
+
+
+def get_content_metadata(request):
+    """
+    Read the content and metadata associated to an HttpRequest.
+
+    Returns:
+        request body (str)
+        request metadata (dict)
+    """
+
+    metadata = {
+        "Content-Type": request.META["CONTENT_TYPE"],
+        "Date": str(timezone.now()),
+        "Content-MD5": hashlib.md5(request.body).hexdigest(),
+        "Content-Length": request.META["CONTENT_LENGTH"],
+    }
+    return request.body, metadata
+
+
+def save_to_file(key, content, metadata=None):
+    """
+    Save the content and metadata to a local file determined by the given key.
+
+    Arguments:
+        key (str): unique file identifier
+        content (str): uploaded file content
+        metadata (dict): json-dumpable data
+    """
+    file_path = get_file_path(key)
+    metadata_path = get_metadata_path(key)
+    if metadata is None:
+        metadata = {}
+
+    safe_save(file_path, content)
+    safe_save(metadata_path, json.dumps(metadata))
```

Comment:
If the first call succeeds and the second call fails, would you want to do any cleanup?